### PR TITLE
fix: honor --exclude when using --as-dockerfile

### PR DESCRIPTION
**Pull request by matejvasek**

### CODE REVIEW

**Comment by matejvasek** [comment]
Maybe this removes need for filtering in tar `type Tar interface` ?

**Comment by matejvasek** [comment]
@coreydaley any idea why the test is failing?

**Comment by coreydaley** [comment]
looking

File: pkg/scm/downloaders/file/download.go [hunk 1]
```diff
@@ -54,9 +55,19 @@ func (f *File) Download(config *api.Config) (*git.SourceInfo, error) {
 		return nil, lerr
 	}
 
+	exclude, err := regexp.Compile(config.ExcludeRegExp)
```
Comment (coreydaley) [inline pkg/scm/downloaders/file/download.go:58]: My first thought is that if `config.ExlcludeRegExp` is empty, the compiled regular expression will match EVERYTHING, so everything will get excluded?
I think you should check it's length first to make sure it's not empty before you compile/use it. And below on line 65 you will probably have to split that into separate checks, one for the `filesToIgnore[path]` and one for the regexp.

Comment (matejvasek) [reply pkg/scm/downloaders/file/download.go:58]: Ah, yes, thanks. Empty string may be indeed problem.
Event with the bug locally most of the integration test passed.
And all unittest passed.

Comment (matejvasek) [reply pkg/scm/downloaders/file/download.go]: The failure looked more like some flake.

**Comment by matejvasek** [comment]
@coreydaley ok, thanks, now test are passing.

**Comment by matejvasek** [comment]
PTAL @coreydaley

File: pkg/scm/downloaders/file/download.go [hunk 2]
```diff
@@ -54,9 +55,27 @@ func (f *File) Download(config *api.Config) (*git.SourceInfo, error) {
...
+			_, ok := filesToIgnore[path]
+			return ok || exclude.MatchString(path)
+		}
+	} else {
+		isIgnored = func(path string) bool {
+			_, ok := filesToIgnore[path]
+			return ok
+		}
+	}
+
```
Comment (coreydaley) [inline pkg/scm/downloaders/file/download.go:75]: how about something like this? (warning, untested)
```suggestion
isIgnored := func(path string) bool {
	if _, ok := filesToIgnore[path]; ok {
			  return ok
	}
	if len(config.ExcludeRegExp) != 0 {
			  return regexp.MustCompile(config.ExcludeRegExp).MatchString(path)
	}
}


```

Comment (matejvasek) [reply pkg/scm/downloaders/file/download.go:75]: @coreydaley  That's quite a pesimisation. It compiles regexp for each check.

If I wanted to optimise code size I would do:
```go
	isIgnored := func(path string) bool {
		_, ok := filesToIgnore[path]
		return ok
	}

	if config.ExcludeRegExp != "" {
		exclude, err := regexp.Compile(config.ExcludeRegExp)
		if err != nil {
			return nil, err
		}
		old := isIgnored
		isIgnored = func(path string) bool {
			return old(path) || exclude.MatchString(path)
		}
	}
```
but that introduces one more indirect call via the closure.

Comment (matejvasek) [reply pkg/scm/downloaders/file/download.go:75]: TBH  I am not huge fan of:
```go
filesToIgnore, lerr := di.GetListOfFilesToIgnore(copySrc)
```
It basically introduces one more traverse on top of the traverse done by `Copy()`.

IMHO the `ignore.DockerIgnorer` struct should have method `Matches(path string) bool` that could we used withing the `Copy()` traversal.

Comment (matejvasek) [reply pkg/scm/downloaders/file/download.go:75]: Not that it made any visible difference on M.2 drives today.

Comment (matejvasek) [reply pkg/scm/downloaders/file/download.go:75]: What I mean:
https://github.com/knative/func/blob/ac575e078f436df13839465d3fb31b2b97ab4a5c/pkg/pipelines/tekton/pipeplines_provider.go#L211-L219

there is even a library like that for .gitignore format.

Comment (matejvasek) [reply pkg/scm/downloaders/file/download.go:75]: >It basically introduces one more traverse on top of the traverse done by `Copy()`.

Actually, maybe it's even worse, it traverse FS for each positive glob in `.s2iignore`.

**Pull request merged**